# Exercise 5.6

We continue to consider the use of a logistic regression model to predict the probability of `default` using `income` and `balance` on the `Default` data set. In particular, we will now compute estimates for the standard errors of the `income` and `balance` logistic regression coefficients in two different ways: (1) using the bootstrap, and (2) using the standard formula for computing the standard errors in the `glm()` function. Do not forget to set a random seed before beginning your analysis.

## (a) Using the `summary()` and `glm()` functions, determine the estimated standard errors for the coefficients associated with `income` and `balance` in a multiple logistic regression model that uses both predictors.

In [ ]:
coef(summary(glm(default ~ income + balance, family="binomial", data=Default)))[, 'Std. Error']

(Intercept)       income      balance 
4.347564e-01 4.985167e-06 2.273731e-04

## (b) Write a function, `boot.fn()`, that takes as input the `Default` data set as well as an index of the observations, and that outputs the coefficient estimates for `income` and `balance` in the multiple logistic regression model.

In [ ]:
boot.fn <- function(data, index=1:nrow(data)) {
    coef(summary(glm(default ~ income + balance, family="binomial", data=data, subset=index)))[-1, 'Estimate']
}
boot.fn(Default)
boot.fn(Default, 1:5000)

income      balance 
2.080898e-05 5.647103e-03

income      balance 
2.885075e-05 5.822416e-03

## (c) Use the `boot()` function together with your `boot.fn()` function to estimate the standard errors of the logistic regression coefficients for `income` and `balance`.

In [ ]:
boot(Default, boot.fn, 1000)


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = Default, statistic = boot.fn, R = 1000)


Bootstrap Statistics :
        original       bias     std. error
t1* 2.080898e-05 1.635325e-07 4.845196e-06
t2* 5.647103e-03 1.842153e-05 2.302679e-04

## (d) Comment on the estimated standard errors obtained using the `glm()` function and using your bootstrap function.

We can also compute the standard error of these bootstrap estimates using the formula (5.8). `glm()` gives the standard errors estimated from the original dataset. Compared with those values, the bootstrap approach also estimates the standard errors well.